mypyc/irbuild/specialize.py

```diff
@@ -135,13 +140,48 @@ def translate_safe_generator_call(
                     + [builder.accept(arg) for arg in expr.args[1:]]),
                 builder.node_type(expr), expr.line, expr.arg_kinds, expr.arg_names)
         else:
+            if callee.fullname == "builtins.tuple":
+                val = tuple_from_generator_helper(builder, expr.args[0])
+                if val is not None:
+                    return val
             return builder.call_refexpr_with_args(
                 expr, callee,
                 ([translate_list_comprehension(builder, expr.args[0])]
                     + [builder.accept(arg) for arg in expr.args[1:]]))
     return None
 
 
+def tuple_from_generator_helper(builder: IRBuilder,
+                                gen: GeneratorExpr) -> Optional[Value]:
+
+    if len(gen.sequences) == 1 and len(gen.condlists[0]) == 0:
+        # Currently we only optimize for simplest generator expression
+        if not is_list_rprimitive(builder.node_type(gen.sequences[0])):
+            return None
+
+        tuple_ops = builder.builder.new_tuple_with_length(builder.accept(gen.sequences[0]),
+                                                          gen.line)
+        index, expr = gen.indices[0], gen.sequences[0]
+
+        def set_tuple_item() -> None:
+            e = builder.accept(gen.left_expr)
+            index_val = builder.accept(index)
+            offset = Integer(1, c_pyssize_t_rprimitive, gen.line)
+            index_pyssize_t = builder.int_op(c_pyssize_t_rprimitive, index_val, offset,
+                                             IntOp.RIGHT_SHIFT, gen.line)
+            index_pyssize_t = builder.int_op(c_pyssize_t_rprimitive, index_pyssize_t, offset,
+                                             IntOp.SUB, gen.line)
+            builder.call_c(new_tuple_set_item_op, [tuple_ops, index_pyssize_t, e], gen.line)
```

Comment:
@JukkaL , I convert `int_rprimitive` to `c_pyssize_t_rprimitive` here. Another implementation would be writing a helper C function in `tuple_ops.c`. Which one is better?

Reply:
minus 1 for correct index

Reply:
`IntOp.RIGHT_SHIFT ` results in a test failure

Reply:
`index_val` on line 168 seems not to be recognized as `int_rprimitive`, how should I fix it?